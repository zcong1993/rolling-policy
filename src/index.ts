export * from './rollingRt'
export * from './rollingCounter'
export * from './rollingPolicy'
export * from './util'
