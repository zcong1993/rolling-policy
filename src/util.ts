export type Hrtime = [number, number]

export const hrtime2nano = (hrtime: Hrtime): number =>
  hrtime[0] * 1e9 + hrtime[1]

export const ms2nano = (t: number): number => t * 1e6

export const sum = (arr: number[]): number => {
  return arr.reduce((prev, acc) => prev + acc, 0)
}
