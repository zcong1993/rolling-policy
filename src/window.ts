export class Bucket {
  private points: number[] = []
  private count: number = 0
  next?: Bucket

  append(val: number) {
    this.points.push(val)
    this.count += 1
  }

  add(offset: number, val: number) {
    this.points[offset] += val
    this.count += 1
  }

  reset() {
    this.points = []
    this.count = 0
  }

  getPoints() {
    return this.points
  }

  getCount() {
    return this.count
  }
}

export class Window {
  private window: Bucket[] = []
  private size: number = 0

  constructor(size: number) {
    this.size = size
    this.window = Array(size)
      .fill(null)
      .map(() => new Bucket())
    for (let i = 0; i < size; i += 1) {
      let nextI = i + 1
      if (nextI === size) {
        nextI = 0
      }
      this.window[i].next = this.window[nextI]
    }
  }

  resetWindow() {
    this.window.forEach(w => w.reset())
  }

  resetBucket(offset: number) {
    this.window[offset].reset()
  }

  append(offset: number, val: number) {
    this.window[offset].append(val)
  }

  add(offset: number, val: number) {
    if (this.window[offset].getCount() === 0) {
      this.window[offset].append(val)
    } else {
      this.window[offset].add(0, val)
    }
  }

  getBucket(offset: number) {
    return this.window[offset]
  }

  getBuckets() {
    return this.window
  }

  getSize() {
    return this.size
  }
}
