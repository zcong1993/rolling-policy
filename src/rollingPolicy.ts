import debug from 'debug'
import cloneDeep from 'clone-deep'
import { Hrtime, hrtime2nano, ms2nano, sum } from './util'

const db = debug('rolling-policy')

export interface Opt {
  window: number // ms, full time window
  windowSize: number // window split bucket size
}

export abstract class RollingPolicy<T> {
  protected windowDuration: number
  protected lastUpdatedAt: Hrtime
  protected startTime: Hrtime
  protected buffer: T[]
  protected opt: Opt

  constructor(c: Opt) {
    this.opt = {
      window: ms2nano(c.window),
      windowSize: c.windowSize,
    }
    this.windowDuration = this.opt.window / this.opt.windowSize
    this.lastUpdatedAt = process.hrtime()
    this.startTime = process.hrtime()

    this.buffer = []
    for (let i = 0; i < c.windowSize; i++) {
      const empty = this.emptyBucketData()
      this.buffer.push(empty)
    }
    db(this)
  }

  add(d: any) {
    this.dropExpired()
    const o = this.offset()
    this.buffer[this.mapIndex(o)] = this.addFunc(
      this.buffer[this.mapIndex(o)],
      d
    )

    db(`index: ${o}, ${this.mapIndex(o)}`)
    this.updatelastUpdatedAt()
  }

  currentBuckets() {
    this.dropExpired()
    return cloneDeep(this.buffer)
  }

  abstract emptyBucketData(): T
  abstract addFunc(old: T, data: any): T

  protected dropExpired() {
    const duration = hrtime2nano(process.hrtime(this.lastUpdatedAt))
    const offset = this.offset()
    const lastOffset = this.lastOffset()
    db(`duration: ${duration}`)
    // reset all
    if (duration > this.opt.window) {
      db('reset all')
      for (let i = 0; i < this.opt.windowSize; i++) {
        this.buffer[i] = this.emptyBucketData()
      }
      return
    }

    const left = offset - this.opt.windowSize
    const lastLeft = lastOffset - this.opt.windowSize
    db(`${lastOffset}, ${offset}, clean: [${lastLeft + 1}, ${left}]`)
    for (let i = lastLeft + 1; i <= left; i++) {
      const mappedIndex = this.mapIndex(i)
      if (mappedIndex >= 0) {
        this.buffer[mappedIndex] = this.emptyBucketData()
      }
    }
  }

  protected updatelastUpdatedAt() {
    this.lastUpdatedAt = process.hrtime()
  }

  protected lastOffset() {
    const o = Math.floor(
      (hrtime2nano(this.lastUpdatedAt) - hrtime2nano(this.startTime)) /
        this.windowDuration
    )
    return o
  }

  protected offset() {
    const diff = hrtime2nano(process.hrtime(this.startTime))
    const o = Math.floor(diff / this.windowDuration)
    return o
  }

  protected mapIndex(i: number) {
    return i % this.opt.windowSize
  }
}
