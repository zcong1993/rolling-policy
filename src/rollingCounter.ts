import { RollingPolicy } from './rollingPolicy'
import { sum } from './util'

type CounterBucket = number[]

export class RollingCounter extends RollingPolicy<CounterBucket> {
  getCounter() {
    return this.currentBuckets().reduce((prev, acc) => prev + sum(acc), 0)
  }

  protected emptyBucketData(): CounterBucket {
    return []
  }

  protected addFunc(old: CounterBucket, data: number): CounterBucket {
    old.push(data)
    return old
  }
}
