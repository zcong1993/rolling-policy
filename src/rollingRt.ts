import { RollingPolicy } from './rollingPolicy'
import { sum } from './util'

interface RtBucket {
  minRt: number
  total: number
  count: number
}

export class RollingRt extends RollingPolicy<RtBucket> {
  average(): RtBucket & { average: number } {
    const buckets = this.currentBuckets()
    const total = sum(buckets.map((b) => b.total))
    const count = sum(buckets.map((b) => b.count))

    return {
      total,
      count,
      minRt: Math.min(...buckets.map((b) => b.minRt)),
      average: count === 0 ? null : total / count,
    }
  }

  protected emptyBucketData(): RtBucket {
    return {
      minRt: Number.MAX_SAFE_INTEGER,
      total: 0,
      count: 0,
    }
  }

  protected addFunc(old: RtBucket, data: number): RtBucket {
    if (data < old.minRt) {
      old.minRt = data
    }
    old.count += 1
    old.total += data
    return old
  }
}
