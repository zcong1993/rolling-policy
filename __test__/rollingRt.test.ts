import { RollingRt } from '../src'

const sleep = (n: number) => new Promise((r) => setTimeout(r, n))

it('should work well', async () => {
  const rc = new RollingRt({ window: 1000, windowSize: 4 })

  expect(rc.average()).toEqual({
    minRt: Number.MAX_SAFE_INTEGER,
    total: 0,
    count: 0,
    average: null,
  })
  for (let i = 0; i < 10; i++) {
    rc.add(1)
    expect(rc.average().count).toBe(Math.min(i + 1, 4))
    expect(rc.average().total).toBe(Math.min(i + 1, 4))
    expect(rc.currentBuckets()).toMatchSnapshot()
    await sleep(250)
  }
})

it('minRt should work well', async () => {
  const rc = new RollingRt({ window: 200, windowSize: 2 })
  for (let i = 0; i < 2; i++) {
    rc.add(10 - i)
    expect(rc.average().minRt).toBe(10 - i)
    await sleep(100)
  }

  await sleep(200)

  for (let i = 0; i < 2; i++) {
    rc.add(10 + i)
    expect(rc.average().minRt).toBe(10)
    await sleep(100)
  }
})
