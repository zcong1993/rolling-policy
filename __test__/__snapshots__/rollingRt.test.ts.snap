// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should work well 1`] = `
Array [
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 0,
    "minRt": 9007199254740991,
    "total": 0,
  },
  Object {
    "count": 0,
    "minRt": 9007199254740991,
    "total": 0,
  },
  Object {
    "count": 0,
    "minRt": 9007199254740991,
    "total": 0,
  },
]
`;

exports[`should work well 2`] = `
Array [
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 0,
    "minRt": 9007199254740991,
    "total": 0,
  },
  Object {
    "count": 0,
    "minRt": 9007199254740991,
    "total": 0,
  },
]
`;

exports[`should work well 3`] = `
Array [
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 0,
    "minRt": 9007199254740991,
    "total": 0,
  },
]
`;

exports[`should work well 4`] = `
Array [
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
]
`;

exports[`should work well 5`] = `
Array [
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
]
`;

exports[`should work well 6`] = `
Array [
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
]
`;

exports[`should work well 7`] = `
Array [
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
]
`;

exports[`should work well 8`] = `
Array [
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
]
`;

exports[`should work well 9`] = `
Array [
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
]
`;

exports[`should work well 10`] = `
Array [
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
  Object {
    "count": 1,
    "minRt": 1,
    "total": 1,
  },
]
`;
